bin/tests/e2e.go

```diff
@@ -0,0 +1,155 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package tests
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+
+	"github.com/magefile/mage/sh"
+	"github.com/pkg/errors"
+)
+
+var projectName = "node_e2e_test"
+var composeFiles = []string{
+	"bin/localnet/docker-compose.yml",
+	"e2e/docker-compose.yml",
+}
+
+type runner struct {
+	compose         func(args ...string) error
+	etherPassphrase string
+}
+
+// TestE2E runs end-to-end tests
+func TestE2E() error {
+	r := newRunner()
+	defer r.cleanup()
+
+	if err := r.setup(); err != nil {
+		return err
+	}
+
+	return r.test()
+}
+
+func (r *runner) setup() error {
+	fmt.Println("initializing geth node")
+	if err := r.compose("run", "geth", "init", "genesis.json"); err != nil {
+		return errors.Wrap(err, "initializing geth node failed!")
+	}
+	fmt.Println("starting other services")
+	if err := r.compose("up", "-d", "broker", "geth"); err != nil {
+		return errors.Wrap(err, "starting other services failed!")
+	}
+	fmt.Println("starting DB")
+	if err := r.compose("up", "-d", "db"); err != nil {
+		return errors.Wrap(err, "starting DB failed!")
+	}
+
+	for {
+		err := r.compose("exec", "-T", "db", "mysqladmin", "ping", "--protocol=TCP", "--silent")
+		if err != nil {
+			fmt.Print(".")
+		} else {
+			fmt.Println()
+			break
+		}
+	}
+
+	fmt.Println("migrating DB")
+	if err := r.compose("run", "--entrypoint", "bin/db-upgrade", "discovery"); err != nil {
+		return errors.Wrap(err, "migrating DB failed!")
+	}
+
+	fmt.Println("starting discovery")
+	if err := r.compose("up", "-d", "discovery"); err != nil {
+		return errors.Wrap(err, "starting discovery failed!")
+	}
+
+	file, err := os.Open("bin/localnet/deployer/local_acc_password.txt")
+	if err != nil {
+		return errors.Wrap(err, "failed to read ether passphrase!")
+	}
+	scanner := bufio.NewScanner(file)
+	if !scanner.Scan() {
+		return errors.New("failed to read ether passphrase!")
+	}
+	r.etherPassphrase = scanner.Text()
+
+	fmt.Println("deploying contracts")
+	err = r.compose("run", "go-runner",
+		"go", "run", "bin/localnet/deployer/deployer.go",
+		"--keystore.directory=bin/localnet/deployer/keystore",
+		"--ether.address=0xa754f0d31411d88e46aed455fa79b9fced122497",
+		"--ether.passphrase", r.etherPassphrase,
+		"--geth.url=http://geth:8545")
+	if err != nil {
+		return errors.Wrap(err, "failed to deploy contracts!")
+	}
+
+	fmt.Println("building app images")
+	if err := r.compose("build"); err != nil {
+		return errors.Wrap(err, "building app images failed!")
+	}
+	fmt.Println("starting app containers")
+	if err := r.compose("up", "-d", "myst-provider", "myst-consumer", "ipify"); err != nil {
+		return errors.Wrap(err, "starting app containers failed!")
```

Comment:
`errors.Wrap()` return nil, if `err == nil`, so you can just do:
```go
err := r.compose("up", "-d", "myst-provider", "myst-consumer", "ipify");
return errors.Wrap(err, "starting app containers failed!")
```

Reply:
Yeah, I can, but it is harder to comprehend, and is inconsistent with all the previous statements. I suppose it could be used for short / one-line functions and such.